### PR TITLE
Updated tests, docs and minor optimization 

### PR DESCRIPTION
**Pull request by rustamwin**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ❌
| Breaks BC?    | ❌
| Tests pass?   | ✔️
| Fixed issues  | -

### CODE REVIEW

File: tests/UrlGeneratorTest.php [hunk 1]
```diff
@@ -18,10 +19,7 @@ private function createUrlGenerator(array $routes): UrlGeneratorInterface
         return $factory($routes, $container);
     }
 
-    /**
-     * @test
-     */
-    public function simpleRouteShouldBeGenerated(): void
+    public function testSimpleRouteShouldBeGenerated(): void
```
Comment (samdark) [inline tests/UrlGeneratorTest.php]:
```suggestion
    public function testSimpleRouteGenerated(): void
```

File: tests/UrlGeneratorTest.php [hunk 2]
```diff
@@ -31,10 +29,7 @@ public function simpleRouteShouldBeGenerated(): void
         $this->assertEquals('/home/index', $url);
     }
 
-    /**
-     * @test
-     */
-    public function routeWithoutNameShouldNotBeFound(): void
+    public function testRouteWithoutNameShouldNotBeFound(): void
```
Comment (samdark) [inline tests/UrlGeneratorTest.php]:
```suggestion
    public function testRouteWithoutNameNotFound(): void
```

File: tests/UrlGeneratorTest.php [hunk 3]
```diff
@@ -47,10 +42,7 @@ public function routeWithoutNameShouldNotBeFound(): void
         $urlGenerator->generate('index');
     }
 
-    /**
-     * @test
-     */
-    public function parametersShouldBeSubstituted(): void
+    public function testParametersShouldBeSubstituted(): void
```
Comment (samdark) [inline tests/UrlGeneratorTest.php]:
```suggestion
    public function testParametersSubstituted(): void
```

File: tests/UrlGeneratorTest.php [hunk 4]
```diff
@@ -60,10 +52,7 @@ public function parametersShouldBeSubstituted(): void
         $this->assertEquals('/view/100/~test#yii', $url);
     }
 
-    /**
-     * @test
-     */
-    public function exceptionShouldBeThrownIfParameterPatternDoesntMatch(): void
+    public function testExceptionShouldBeThrownIfParameterPatternDoesntMatch(): void
```
Comment (samdark) [inline tests/UrlGeneratorTest.php]:
```suggestion
    public function testExceptionThrownIfParameterPatternDoesntMatch(): void
```

File: tests/UrlGeneratorTest.php [hunk 5]
```diff
@@ -74,10 +63,7 @@ public function exceptionShouldBeThrownIfParameterPatternDoesntMatch(): void
         $urlGenerator->generate('view', ['id' => null]);
     }
 
-    /**
-     * @test
-     */
-    public function exceptionShouldBeThrownIfAnyParameterIsMissing(): void
+    public function testExceptionShouldBeThrownIfAnyParameterIsMissing(): void
```
Comment (samdark) [inline tests/UrlGeneratorTest.php]:
```suggestion
    public function testExceptionThrownIfAnyParameterIsMissing(): void
```

File: tests/UrlGeneratorTest.php [hunk 6]
```diff
@@ -88,16 +74,13 @@ public function exceptionShouldBeThrownIfAnyParameterIsMissing(): void
         $urlGenerator->generate('view', ['id' => 123]);
     }
 
-    /**
-     * @test
-     */
-    public function groupPrefixShouldBeAppended(): void
+    public function testGroupPrefixShouldBeAppended(): void
```
Comment (samdark) [inline tests/UrlGeneratorTest.php]:
```suggestion
    public function testGroupPrefixAppended(): void
```

File: tests/UrlGeneratorTest.php [hunk 7]
```diff
@@ -108,10 +91,7 @@ public function groupPrefixShouldBeAppended(): void
         $this->assertEquals('/api/post/42', $url);
     }
 
-    /**
-     * @test
-     */
-    public function defaultSholdNotBeUsedForOptionalParameter(): void
+    public function testDefaultShouldNotBeUsedForOptionalParameter(): void
```
Comment (samdark) [inline tests/UrlGeneratorTest.php]:
```suggestion
    public function testDefaultNotUsedForOptionalParameter(): void
```

File: tests/UrlGeneratorTest.php [hunk 8]
```diff
@@ -123,10 +103,7 @@ public function defaultSholdNotBeUsedForOptionalParameter(): void
         $this->assertEquals('/', $url);
     }
 
-    /**
-     * @test
-     */
-    public function valueShouldBeUsedForOptionalParameter(): void
+    public function testValueShouldBeUsedForOptionalParameter(): void
```
Comment (samdark) [inline tests/UrlGeneratorTest.php]:
```suggestion
    public function testValueUsedForOptionalParameter(): void
```

File: tests/UrlGeneratorTest.php [hunk 9]
```diff
@@ -138,10 +115,7 @@ public function valueShouldBeUsedForOptionalParameter(): void
         $this->assertEquals('/test', $url);
     }
 
-    /**
-     * @test
-     */
-    public function defaultShouldNotBeUsedForRequiredParameter(): void
+    public function testDefaultShouldNotBeUsedForRequiredParameter(): void
```
Comment (samdark) [inline tests/UrlGeneratorTest.php]:
```suggestion
    public function testDefaultNotUsedForRequiredParameter(): void
```

File: tests/UrlMatcherTest.php [hunk 10]
```diff
@@ -18,10 +18,7 @@ private function createUrlMatcher(array $routes): UrlMatcherInterface
         return $factory($routes, $container);
     }
 
-    /**
-     * @test
-     */
-    public function defaultsShouldBeInResult(): void
+    public function testDefaultsShouldBeInResult(): void
```
Comment (samdark) [inline tests/UrlMatcherTest.php]:
```suggestion
    public function testDefaultsAreInResult(): void
```

File: src/FastRoute.php [hunk 11]
```diff
@@ -437,7 +437,7 @@ function (RouteCollector $r) use ($group) {
                     $modifiedItem = $item->pattern($group->getPrefix() . $item->getPattern());
                     $groupMiddlewares = $group->getMiddlewares();
 
-                    for (end($groupMiddlewares); key($groupMiddlewares) !== null; prev($groupMiddlewares)) {
+                    for (end($groupMiddlewares), $loopsMax = key($groupMiddlewares); $loopsMax !== null; prev($groupMiddlewares)) {
```
Comment (yiiliveext) [inline src/FastRoute.php]: You are wrong it will not work.

Comment (samdark) [reply src/FastRoute.php]: I confirmed it.

**Comment by samdark** [comment]
Thanks!

**Pull request merged**